firebase-perf/src/main/java/com/google/firebase/perf/application/FrameMetricsRecorder.java

```diff
@@ -0,0 +1,182 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.perf.application;
+
+import android.app.Activity;
+import android.util.SparseIntArray;
+import androidx.core.app.FrameMetricsAggregator;
+import com.google.android.gms.common.util.VisibleForTesting;
+import com.google.firebase.perf.logging.AndroidLogger;
+import com.google.firebase.perf.metrics.FrameMetricsCalculator.PerfFrameMetrics;
+import com.google.firebase.perf.util.Constants;
+import com.google.firebase.perf.util.Optional;
+import java.util.WeakHashMap;
+
+/**
+ * Provides FrameMetrics data from an Activity's Window. Encapsulates FrameMetricsAggregator.
+ *
+ * <p>Note: each recorder holds a reference to an Activity, so it is very important to dereference
+ * each recorder before the associated Activity is destroyed
+ */
+public class FrameMetricsRecorder {
+  private static final AndroidLogger logger = AndroidLogger.getInstance();
+  private final Activity activity;
+  private final FrameMetricsAggregator fma;
+  private final WeakHashMap<Object, PerfFrameMetrics> subTraceMap;
+  private boolean isRecording = false;
+
+  /**
+   * Creates a recorder for a specific activity
+   *
+   * @param activity the activity that the recorder is collecting data from
+   */
+  public FrameMetricsRecorder(Activity activity) {
+    this(activity, new FrameMetricsAggregator(), new WeakHashMap<>());
+  }
+
+  @VisibleForTesting
+  FrameMetricsRecorder(
+      Activity activity,
+      FrameMetricsAggregator frameMetricsAggregator,
+      WeakHashMap<Object, PerfFrameMetrics> subTraceMap) {
+    this.activity = activity;
+    this.fma = frameMetricsAggregator;
+    this.subTraceMap = subTraceMap;
+  }
+
+  /** Starts recording FrameMetrics for the activity window */
+  public void start() {
+    if (isRecording) {
+      logger.error(
+          "FrameMetricsAggregator is already recording %s", activity.getClass().getSimpleName());
+      return;
+    }
+    fma.add(activity);
+    isRecording = true;
+  }
+
+  /**
+   * Stops recording FrameMetrics for the activity window
+   *
+   * @return FrameMetrics accumulated during the current recording
+   */
+  public Optional<PerfFrameMetrics> stop() {
+    if (!isRecording) {
+      logger.error("Cannot stop because no recording was started");
+      return Optional.absent();
+    }
+    Optional<PerfFrameMetrics> data = this.snapshot();
+    try {
+      fma.remove(activity);
+    } catch (IllegalArgumentException err) {
+      logger.debug("View not hardware accelerated. Unable to collect FrameMetrics. %s", err);
+    }
+    fma.reset();
+    isRecording = false;
+    return data;
+  }
+
+  /**
+   * Starts a sub-trace in the current recording
```

Comment:
I'm not really sure what a sub-trace is and why it needs to be recorded in this class.

Reply:
`FrameMetricsAggregator` collects frame-time from an Activity's Window, for every frame rendered for that Window. Fragment screen trace is only an interval within that same frame-time data from the Window. `sub-trace` may be a bad name but I couldn't think of another word to describe this relationship, open to more suggestions.

This is also the reason I made the parameter `Object`, because from `FrameMetricsAggregator`'s perspective, all it does and all it can do is collect frame-times for an Activity's Window. Fragment screen trace is a "UI state" we arbitrarily added to tag an interval within that frame-time recording. `FrameMetricsAggregator` shouldn't care if you use anything like `Fragment` or other UI objects, all it cares about is that you want a sub-interval of frametimes from time A to time B, and you want to give it a name "Fragment A".

Reply:
Discussed offline. There are 2 main options:

1. Using "fragment" instead of sub-trace throughout this class for now and update it when we have a second use case.
2. Keep it generic, but then we need to add some more javadoc to explain the generic concepts we are using.

Reply:
Using "fragment" for now. Will generalize it in the future if needed.